runtime/expr/agg/map.go

```diff
@@ -0,0 +1,98 @@
+package agg
+
+import (
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/zcode"
+)
+
+type Map struct {
+	entries map[string]mapEntry
+	values  []zed.Value
+	size    int
+	scratch []byte
+}
+
+func newMap() *Map {
+	return &Map{entries: make(map[string]mapEntry)}
+}
+
+var _ Function = (*Collect)(nil)
+
+type mapEntry struct {
+	key *zed.Value
+	val *zed.Value
+}
+
+func (m *Map) Consume(val *zed.Value) {
+	if val.IsNull() {
+		return
+	}
+	mtyp, ok := zed.TypeUnder(val.Type).(*zed.TypeMap)
+	if !ok {
+		return
+	}
+	it := val.Bytes.Iter()
+	for !it.Done() {
+		ktyp, kb := untagUnionValue(mtyp.KeyType, it.Next())
+		vtyp, vb := untagUnionValue(mtyp.ValType, it.Next())
```

Comment:
Use expr.ValueUnder instead of untagUnionValue.

Reply:
This will strip named types, is that something we want to do here?

Reply:
You want to removed the named type if its a named type of a union (I should fix that in `untagUnionValue`).

Reply:
> This will strip named types, is that something we want to do here?

Good point!

I think that means we want something like this:
```go
// valueUnder is like expr.ValueUnder but it preserves non-union named types.
func valueUnder(typ zed.Type, b zcode.Bytes) *zed.Value {
	val := zed.NewValue(typ, b)
	if _, ok := zed.TypeUnder(typ).(*zed.TypeUnion); !ok {
		return val
	}
	return expr.ValueUnder(val)
}
```

Reply:
Ugh, I can't use expr.ValueUnder since it creates an import cycle.

Reply:
Maybe make it a method on zed.Value?

Reply:
Interesting